crates/forge/bin/cmd/bind_json.rs

```diff
@@ -0,0 +1,530 @@
+        });
+
+        let output = with_compilation_reporter(false, || {
+            ProjectCompiler::with_sources(&project, sources)?.compile()
+        })?;
+
+        if output.has_compiler_errors() {
+            eyre::bail!("{output}");
+        }
+
+        // Collect ASTs by getting them from sources and converting into strongly typed
+        // `SourceUnit`s. Also strips root from paths.
+        let asts = output
+            .into_output()
+            .sources
+            .into_iter()
+            .filter_map(|(path, mut sources)| Some((path, sources.swap_remove(0).source_file.ast?)))
+            .map(|(path, ast)| {
+                Ok((
+                    path.strip_prefix(project.root()).unwrap_or(&path).to_path_buf(),
+                    serde_json::from_str::<SourceUnit>(&serde_json::to_string(&ast)?)?,
+                ))
+            })
+            .collect::<Result<BTreeMap<_, _>>>()?;
+
+        Ok(CompiledState { asts, target_path, config, project })
+    }
+}
+
+#[derive(Debug, Clone)]
+struct CompiledState {
+    asts: BTreeMap<PathBuf, SourceUnit>,
+    target_path: PathBuf,
+    config: Config,
+    project: Project,
+}
+
+impl CompiledState {
+    fn find_structs(self) -> Result<StructsState> {
+        let Self { asts, target_path, config, project } = self;
+
+        // construct mapping (file, id) -> (struct definition, optional parent contract name)
+        let structs = asts
+            .iter()
+            .flat_map(|(path, ast)| {
+                let mut structs = Vec::new();
+                // we walk AST directly instead of using visitors because we need to distinguish
+                // between file-level and contract-level struct definitions
+                for node in &ast.nodes {
+                    match node {
+                        SourceUnitPart::StructDefinition(def) => {
+                            structs.push((def, None));
+                        }
+                        SourceUnitPart::ContractDefinition(contract) => {
+                            for node in &contract.nodes {
+                                if let ContractDefinitionPart::StructDefinition(def) = node {
+                                    structs.push((def, Some(contract.name.clone())));
+                                }
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                structs.into_iter().map(|(def, parent)| ((path.as_path(), def.id), (def, parent)))
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        // Resolver for EIP712 schemas
+        let resolver = Resolver::new(&asts);
+
+        let mut structs_to_write = Vec::new();
+
+        let include =
+            config.bind_json.include.into_iter().map(GlobMatcher::new).collect::<Vec<_>>();
+        let exclude =
+            config.bind_json.exclude.into_iter().map(GlobMatcher::new).collect::<Vec<_>>();
```

Comment:
can we not put the struct in the config directly?

Reply:
yep, just needed manual `PartialEq` a227ed8